fix path segment parsing when suffixed with query

yisraelU opened the pull request:
In the scenario of a constant query (and perhaps some other cases) parsing of path segments fail.
`http://yourhost/my/uri/{foo}?hello=world` should parse a label foo, which it currently does not

yisraelU committed:
fix path segment parsing when suffixed with query

daddykotex commented on modules/core/src/smithy4s/http/internals/package.scala:
```diff
@@ -63,12 +63,17 @@ package object internals {
-    else if (str.startsWith("{") && str.endsWith("}"))
-      Some(PathSegment.label(str.substring(1, str.length() - 1)))
-    else Some(PathSegment.static(str))
+    Option(str).map { str =>
```
I think you got rid of a `.isEmpty()` here

you may want to reintroduce it as a `.filter(_.nonEmpty)` on the `Option`

yisraelU replied:
ah yes , was thinking of Null , will update

yisraelU committed:
add back filtering for non emptiness on string

kubukoz commented on modules/core/test/src/smithy4s/http/internals/PathSpec.scala:
```diff
@@ -46,7 +46,7 @@ class PathSpec() extends munit.FunSuite {
 
   test("Parse path pattern into path segments") {
-    val result = pathSegments("/{head}/foo/{tail+}")
+    val result = pathSegments("/{head}/foo/{tail+}?hello=world")
```
imo this deserves a separate test case

Baccata replied:
I concur

yisraelU committed:
adds separate test for parsing path params from path with query params

yisraelU committed:
Changes:
PathSegment implementation returns just the path, minus any query parameters
A  staticQueryParams method has been added  HttpEndpoint,
the Http4s client calls that method to retrieve any static query params and pass them on

Tests: metadata.smithy has been modified and PathSpec includes tests for testing the HttpEndpoint functionality

yisraelU committed:
formatting

Baccata commented on modules/core/src/smithy4s/http/HttpEndpoint.scala:
```diff
@@ -26,6 +26,9 @@ trait HttpEndpoint[I] {
   // Returns a path template as a list of segments, which can be constant strings or placeholders.
   def path: List[PathSegment]
+
+  // Returns a map of static query parameters that are found in the uri of Http hint, there cannot be multiple of the same key !.
```
even though that's the case at the smithy level, I'd be more comfortable with modelling it in this  interface as a multimap, to future-proof it in case it ever changes.

yisraelU replied:
fair enough , Ill revert

Baccata commented on modules/http4s/src/smithy4s/http4s/internals/SmithyHttp4sClientEndpoint.scala:
```diff
@@ -114,8 +114,10 @@ private[http4s] class SmithyHttp4sClientEndpointImpl[F[_], Op[_, _, _, _, _], I,
+    val staticQueries = httpEndpoint.staticQueryParams
     val uri = baseUri
       .copy(path = baseUri.path.addSegments(path.map(Uri.Path.Segment(_))))
+      .withQueryParams(staticQueries)
```
👍

yisraelU committed:
revert staticQueryParams to Map[String,Seq[String]] for future proofing sake

yisraelU committed:
fix test for new statiq query params interface

Baccata commented on modules/core/src/smithy4s/http/internals/package.scala:
```diff
@@ -56,19 +56,43 @@ package object internals {
-      .filterNot(_.isEmpty())
-      .traverse(fromToString(_))
+      .split('?')
+      .headOption
```
nitpick : here it's safe to call `.head` instead of `.headOption`, that'll save us an allocation

Baccata commented on modules/core/src/smithy4s/http/internals/package.scala:
```diff
@@ -56,19 +56,43 @@ package object internals {
-    else if (str.startsWith("{") && str.endsWith("}"))
-      Some(PathSegment.label(str.substring(1, str.length() - 1)))
-    else Some(PathSegment.static(str))
+    Option(str).filter(_.nonEmpty).map { str =>
```
nitpick : here too : I'd rather have a `if (str == null || str.isEmpty)` to save an allocation

yisraelU committed:
removed extra allocations

Baccata reviewed (approved):
Good stuff !
